File: AdventOfCode2025.FSharp/Day01.fs
module AdventOfCode2025.FSharp

open System.IO

type Direction =
    | Left
    | Right

type Instruction = {
    Direction: Direction
    Steps: int
}

module Instruction =
    let parse (line: string) : Instruction =
        let dir =
            match line[0] with
            | 'L' -> Left
            | 'R' -> Right
            | c   -> failwithf $"Unexpected direction '%c{c}' in line: %s{line}"
            
        { Direction = dir; Steps = line[1..] |> int }

module Day01Logic =

    let turnDial (currentPos: int) (direction: Direction) : int =
        let step =
            match direction with
            | Left  -> -1
            | Right ->  1
            
        (currentPos + step + 100) % 100
        
    let processLine (currentPos: int) (line: string) : int * int =
        let instruction =
            Instruction.parse line

        let folder (cur, pass) _ =
            let next = turnDial cur instruction.Direction
            let pass' = if next = 0 then pass + 1 else pass
            next, pass'

        [ 1 .. instruction.Steps ]
        |> List.fold folder (currentPos, 0)
    
    let solve (lines: seq<string>) =
        lines
        |> Seq.fold
            (fun (dialPos, exactCount, passCount) line ->
                let dialPosAfter, passDelta = processLine dialPos line
                let exactCountDelta = if dialPosAfter = 0 then 1 else 0
                dialPosAfter, exactCount + exactCountDelta, passCount + passDelta
            )
            (50, 0, 0)

type Day01() =
    interface ISolver<Day01> with
        static member DayNumber = 1;
        member _.Solve(inputFile: FileInfo) : struct (string * string) =
            let _, exact, pass = File.ReadLines inputFile.FullName |> Day01Logic.solve
            struct (string exact, string pass)

File: AdventOfCode2025.CSharp/Day01.cs
namespace AdventOfCode2025.CSharp;

public sealed class Day01 : ISolver<Day01>
{
    public static int DayNumber => 1;
    
    public (object? PartOne, object? PartTwo) Solve(FileInfo inputFile)
    {
        var dial = new Dial();
        var exactlyZeroCount = 0;
        var passThroughZeroCount = 0;
        
        foreach (var line in File.ReadLines(inputFile.FullName))
        {
            Action turnDial = line[0] is 'L' ? dial.TurnLeft : dial.TurnRight;
            var ticks = int.Parse(line[1..]);

            for (var i = 0; i < ticks; i++)
            {
                turnDial();
                if (dial.CurrentValue == 0) passThroughZeroCount++;
            }

            if (dial.CurrentValue == 0) exactlyZeroCount++;
        }
        
        return (exactlyZeroCount, passThroughZeroCount);
    }

    public sealed class Dial
    {
        public int CurrentValue { get; private set; } = 50;

        public void TurnLeft() => CurrentValue = (CurrentValue + 99) % 100;
        public void TurnRight() => CurrentValue = (CurrentValue + 101) % 100;
    }
}